packages/user/src/utils.ts

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { URLExt } from '@jupyterlab/coreutils';
+import { ServerConnection } from '@jupyterlab/services';
+
+/**
+ * Call the API extension
+ *
+ * @param endPoint API REST end point for the extension
+ * @param init Initial values for the request
+ * @returns The response body interpreted as JSON
+ */
+export async function requestAPI<T>(
+  endPoint = '',
+  init: RequestInit = {}
+): Promise<T> {
+  const settings = ServerConnection.makeSettings();
+  const requestUrl = URLExt.join(settings.baseUrl, endPoint);
+
+  let response: Response;
+  try {
+    response = await ServerConnection.makeRequest(requestUrl, init, settings);
+  } catch (error) {
+    throw new ServerConnection.NetworkError(error);
+  }
+
+  let data: any = await response.text();
+
+  if (data.length > 0) {
+    try {
+      data = JSON.parse(data);
+    } catch (error) {
+      console.log('Not a JSON response body.', response);
+    }
+  }
+
+  if (!response.ok) {
+    throw new ServerConnection.ResponseError(response, data.message || data);
+  }
+
+  return data;
+}
+
+// From https://en.wikipedia.org/wiki/Moons_of_Jupiter
+export const moonsOfJupyter = [
+  'Metis',
+  'Adrastea',
+  'Amalthea',
+  'Thebe',
+  'Io',
+  'Europa',
+  'Ganymede',
+  'Callisto',
+  'Themisto',
+  'Leda',
+  'Ersa',
+  'Pandia',
+  'Himalia',
+  'Lysithea',
+  'Elara',
+  'Dia',
+  'Carpo',
+  'Valetudo',
+  'Euporie',
+  'Eupheme',
+  // 'S/2003 J 18',
+  // 'S/2010 J 2',
+  'Helike',
+  // 'S/2003 J 16',
+  // 'S/2003 J 2',
+  'Euanthe',
+  // 'S/2017 J 7',
+  'Hermippe',
+  'Praxidike',
+  'Thyone',
+  'Thelxinoe',
+  // 'S/2017 J 3',
+  'Ananke',
+  'Mneme',
+  // 'S/2016 J 1',
+  'Orthosie',
+  'Harpalyke',
+  'Iocaste',
+  // 'S/2017 J 9',
+  // 'S/2003 J 12',
+  // 'S/2003 J 4',
+  'Erinome',
+  'Aitne',
+  'Herse',
+  'Taygete',
+  // 'S/2017 J 2',
+  // 'S/2017 J 6',
+  'Eukelade',
+  'Carme',
+  // 'S/2003 J 19',
+  'Isonoe',
+  // 'S/2003 J 10',
+  'Autonoe',
+  'Philophrosyne',
+  'Cyllene',
+  'Pasithee',
+  // 'S/2010 J 1',
+  'Pasiphae',
+  'Sponde',
+  // 'S/2017 J 8',
+  'Eurydome',
+  // 'S/2017 J 5',
+  'Kalyke',
+  'Hegemone',
+  'Kale',
+  'Kallichore',
+  // 'S/2011 J 1',
+  // 'S/2017 J 1',
+  'Chaldene',
+  'Arche',
+  'Eirene',
+  'Kore',
+  // 'S/2011 J 2',
+  // 'S/2003 J 9',
+  'Megaclite',
+  'Aoede',
+  // 'S/2003 J 23',
+  'Callirrhoe',
+  'Sinope'
+];
+
+/**
+ * Get a random user-name based on the moons of Jupyter.
+ * This function returns names like "Anonymous Io" or "Anonymous Metis".
+ */
+export const getAnonymousUserName = (): string =>
+  'Anonymous ' +
+  moonsOfJupyter[Math.floor(Math.random() * moonsOfJupyter.length)];
+
+/**
+ * Predefined colors for users
+ */
+export const userColors = [
```

Comment:
Can we instead use a categorical color scheme that has been well tested for usability, constrast, etc. Something like category20 is a good starting point. We are also using this in other parts of JupyterLab (coloring autocompletion results).

https://vega.github.io/vega/docs/schemes/#category20

Reply:
Yes, choose the colors, and I'll add them.
The colors we are using are the ones @dmonad uses in YJS, and they were introduced in the RTC PR.

Reply:
Here are the category 20 colors (as a concatenated string, I can separate these): https://github.com/vega/vega/blob/master/packages/vega-scale/src/palettes.js#L67

Reply:
The current colors we have were already checked for contrast: https://github.com/jupyterlab/jupyterlab/pull/10411#discussion_r651960093 and I haven't seen any complaints (does not mean those are the best they could be). @ellisonbg is there any colour on the current list that does not work well for you?

Vega's `category20` colors have excellent contrast against black text, but of the 20 colors from only 7 pass the contrast check for WCAG AA for large text when used as background for **white** text, which is the case when showing the user cursor; showing it with black text would be distracting (and depart from the convention adopted by Google Docs, Deepnote and many others). Many in `category20` fail badly (contrast ratio < 2.5):

```js
[
  '#1f77b4',   # 4.82
  '#aec7e8',   # 1.73 fail
  '#ff7f0e',   # 2.53 fail
  '#ffbb78',   # 1.66 fail
  '#2ca02c',   # 3.40
  '#98df8a',   # 1.58 fail
  '#d62728',   # 5.02
  '#ff9896',   # 2.06 fail
  '#9467bd',   # 4.25
  '#c5b0d5',   # 1.99 fail
  '#8c564b',   # 5.91
  '#c49c94',   # 2.46 fail
  '#e377c2',   # 2.73 fail
  '#f7b6d2',   # 1.66 fail
  '#7f7f7f',   # 4.00
  '#c7c7c7',   # 1.69 fail
  '#bcbd22',   # 2.01 fail
  '#dbdb8d',   # 1.44 fail
  '#17becf',   # 2.25 fail
  '#9edae5'    # 1.54
]
```

`category20b` would be a better starting point - only 8 failing contrast with white text:

```js
[
  '#393b79',   # 10.11
  '#5254a3',   # 6.65
  '#6b6ecf',   # 4.41
  '#9c9ede',   # 2.51 fail
  '#637939',   # 4.85
  '#8ca252',   # 2.83 fail
  '#b5cf6b',   # 1.73 fail
  '#cedb9c',   # 1.47 fail
  '#8c6d31',   # 4.83
  '#bd9e39',   # 2.58 fail
  '#e7ba52',   # 1.81 fail
  '#e7cb94',   # 1.57 fail
  '#843c39',   # 7.81
  '#ad494a',   # 5.48
  '#d6616b',   # 3.63
  '#e7969c',   # 2.27 fail
  '#7b4173',   # 7.37
  '#a55194',   # 4.99
  '#ce6dbd',   # 3.2
  '#de9ed6'    # 2.1 fail
]
```

There is also `category20c` is not good either:


```js
[
  '#3182bd',   # 4.15
  '#6baed6',   # 2.42 fail
  '#9ecae1',   # 1.75 fail
  '#c6dbef',   # 1.42 fail
  '#e6550d',   # 3.70
  '#fd8d3c',   # 2.31 fail
  '#fdae6b',   # 1.83 fail
  '#fdd0a2',   # 1.42 fail
  '#31a354',   # 3.23
  '#74c476',   # 2.12 fail
  '#a1d99b',   # 1.62 fail
  '#c7e9c0',   # 1.32 fail
  '#756bb1',   # 4.67
  '#9e9ac8',   # 2.65 fail
  '#bcbddc',   # 1.83 fail
  '#dadaeb',   # 1.38 fail
  '#636363',   # 6.00
  '#969696',   # 2.95 fail
  '#bdbdbd',   # 1.87 fail
  '#d9d9d9'    # 1.41 fail
]
```

The truth is there is only so many colors you can show with good contrast against text and with good contrast against each other. From experience, it is much less than 20.

You can use if you want to test another pallette: https://webaim.org/resources/contrastchecker/

Reply:
Actually, I was thinking a bit more about the color scheme. I think it is important for the user colors to be something that are in the theme, so that themes an modify the colors used.

Reply:
Yep, you are right. The categoryX color sequences fair pretty poorly as a background upon which to render text. At the same time, I think there are two contrast dimensions this sequence needs to cover: 1) the contrast between the colors, so users are distinguishable by color, 2) the contrast of text (white or black) on the color.

My initial thought is the generate a sequence using something like the following: 

* Work in HCL colorspace, and pick a single value for the Luminance. Because the Luminance is the primary dimension that determines the contrast against white/black text, we should pick a L that gives the needed WGAC constrast numbers.
* Pick a mind-ranged Chrome, to give enough color to give a strong visual effect, but not so high as to run into colors that our out of the RGB gamut.
* Then generate the sequence by rotating through the huge values.

I have used the R [colorspace](https://cran.r-project.org/web/packages/colorspace/vignettes/colorspace.html) library before and its [Python port](https://python-colorspace.readthedocs.io/en/stable/hclcolorspace.html).

One question: how were the current colors generated.

Reply:
We could add a categorical color scale to the theme, so that we can override it in custom themes?

Reply:
> 1) the contrast between the colors, so users are distinguishable by color, 2) the contrast of text (white or black) on the color.

and also 3) contrast of the cursor buble with the background; this can be worked around by adding a thin 1-px (or even sub-pixel) outline for dark themes, but yes ideally we would have different variations per theme, but possibly we would not want to change the colours completely (only their Luminance) so that users can still refer to their colours regardless of the theme ("can you see my green cursor?")